shuffle-plugin/src/main/scala/com/nvidia/spark/rapids/shuffle/ucx/UCX.scala

```diff
@@ -345,6 +362,258 @@ class UCX(executor: BlockManagerId, rapidsConf: RapidsConf) extends AutoCloseabl
     })
   }
 
+  /**
+   * This trait and next two implementations represent the mapping between an Active Message Id
+   * and the callback that should be triggered when a message is received.
+   *
+   * There are two types of Active Messages we care about: requests and responses.
+   *
+   * For requests:
+   *   - `activeMessageId` for requests is the value of the `RequestType` enum, and it is
+   *   set once when the `RapidsShuffleServer` is initialized, and no new request handlers
+   *   are established.
+   *
+   *   - The Active Message header is handed to the request handler, via the transaction.
+   *   The request handler needs to echo the header back for the response handler on the
+   *   other side of the request.
+   *
+   *   - On a request, a callback is instantiated using `requestCallbackGen`, which creates
+   *   a transaction each time. This is one way to handle several requests inbound to a server.
+   *
+   * For responses:
+   *   - `activeMessageId` for responses is the value of the `RequestType` enum with an extra
+   *   bit flipped (see `UCXConnection.composeResponseAmId`). These are also set once, as requests
+   *   are sent out.
+   *
+   *   - The Active Message header is used to pick the correct callback to call. In this case
+   *   there could be several expected responses, for a single response `activeMessageId`, so the
+   *   server echoes back our header so we can invoke the correct response callback.
+   *
+   *   - Each response received at the response activeMessageId, will be demuxed using the header:
+   *   responseActiveMessageId1 -> [callbackForHeader1, callbackForHeader2, ..., callbackForHeaderN]
+   */
+  trait ActiveMessageRegistration {
+    val activeMessageId: Int
+    def getCallback(header: Long): UCXAmCallback
+  }
+
+  class RequestActiveMessageRegistration(override val activeMessageId: Int)
+      extends ActiveMessageRegistration {
+    private var requestCallbackGen: () => UCXAmCallback = null
```

Comment:
Why isn't this a constructor argument?  There's only one place that sets it, and that's immediately after constructing the instance.  Is there a valid reason to have an instance where this remains null for any significant duration after construction?

Reply:
Making the change, that will be cleaner for sure.

Reply:
done.

Reply:
whoops trigger happy.. one sec